client/asset/btc/btc.go

```diff
@@ -1747,336 +1733,206 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
-// registered result chan.
-// Also checks mempool for potential redemptions if spending info is not found
-// for any of these contracts in the specified block range.
-func (btc *ExchangeWallet) findRedemptionsInBlockRange(startBlock, endBlock *block, contractOutpoints []outPoint) {
-	contractsCount := len(contractOutpoints)
-	btc.log.Debugf("finding redemptions for %d contracts in blocks %d - %d",
-		contractsCount, startBlock.height, endBlock.height)
+	iHash, iHeight := lowestBlock()
 
-	nextBlockHash := startBlock.hash
-	var lastScannedBlockHeight int64
-	var totalFound, totalCanceled int
-
-rangeBlocks:
-	for nextBlockHash != "" && lastScannedBlockHeight < endBlock.height {
-		blk, err := btc.getVerboseBlockTxs(nextBlockHash)
+	// If a startBlock is provided at a higher height, use that as the starting
+	// point.
+	if startBlock != nil {
+		h, err := btc.node.getBlockHeight(startBlock)
 		if err != nil {
-			// Redemption search for this set of contracts is compromised. Notify
-			// the redemption finder(s) of this fatal error and cancel redemption
-			// search for these contracts. The redemption finder(s) may re-call
-			// btc.FindRedemption to restart find redemption attempts for any of
-			// these contracts.
-			err = fmt.Errorf("error fetching verbose block %s: %w", nextBlockHash, err)
-			btc.fatalFindRedemptionsError(err, contractOutpoints)
+			btc.log.Errorf("tryRedemptionRequests getBlockHeight error: %v", err)
 			return
 		}
-		scanPoint := fmt.Sprintf("block %d", blk.Height)
-		lastScannedBlockHeight = int64(blk.Height)
-		for t := range blk.Tx {
-			tx := &blk.Tx[t]
-			found, canceled := btc.findRedemptionsInTx(scanPoint, tx, contractOutpoints)
-			totalFound += found
-			totalCanceled += canceled
-			if totalFound+totalCanceled == contractsCount {
-				break rangeBlocks
-			}
+		// If the startBlock height is lower than the lowest swap block height,
+		// ignore the startBlock.
+		if h > iHeight {
+			iHeight = h
+			iHash = startBlock
 		}
-		nextBlockHash = blk.NextHash
 	}
 
-	btc.log.Debugf("%d redemptions found, %d canceled out of %d contracts in blocks %d - %d",
-		totalFound, totalCanceled, contractsCount, startBlock.height, lastScannedBlockHeight)
-
-	// Search for redemptions in mempool if there are yet unredeemed
-	// contracts after searching this block range.
-	pendingContractsCount := contractsCount - totalFound - totalCanceled
-	if pendingContractsCount > 0 {
-		btc.findRedemptionMtx.RLock()
-		pendingContracts := make([]outPoint, 0, pendingContractsCount)
-		for _, contractOutpoint := range contractOutpoints {
-			if _, pending := btc.findRedemptionQueue[contractOutpoint]; pending {
-				pendingContracts = append(pendingContracts, contractOutpoint)
+	for iHeight <= tipHeight {
+		validReqs := make(map[outPoint]*findRedemptionReq, len(undiscovered))
+		for outPt, req := range undiscovered {
+			if iHeight >= req.blockHeight {
+				validReqs[outPt] = req
 			}
```

Comment:
I want to recommend checking the findRedemptionQueue at this point to confirm that the undiscovered contract has NOT been discovered by some other search op. E.g. the following sequence of events happen:
- the initial search op for a contract is ongoing 
- a new block is mined and another redemption search is triggered and the redemption is found in the new block
- the initial search keeps trying to find the redemption in blocks between the contract block and the newly mined block